### PR TITLE
fix(controller): Updates to add condition to workflow status. Fixes #2421

### LINKED ISSUE #2421: Add `Conditions` to workflow status
# Summary

Add `Conditions` to workflow status.

# Motivation

From the [kubernetes api conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties):

> Some resources in the v1 API contain fields called phase, and associated message, reason, and other status fields. The pattern of using phase is deprecated. Newer API types should use conditions instead.

This would facilitate using e.g. `kubectl wait` to wait for workflows:

`kubectl wait --for=condition=complete workflow/myworkflow`

The `argo wait` command UI is very nice, not sure if it's possible to get any similar experience with `kubectl wait`. But where `kubectl wait` is handy, is if you are trying to wait on a data drive list of resources, some of which may be workflows, some jobs, or any other type of resource for certain conditions, which our team is looking to do.

# Proposal

Add a `Conditions` field to workflow status, which includes at least a `completed` or similar condition, and whatever other conditions may make sense. Deprecate the `phase` etc. fields in status which are superceded by `Conditions`.

---

**Message from the maintainers**:

If you wish to see this enhancement implemented please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by NikeNano**
This PR intends to fix the requested features in #2421. 

The following changes have been made:

- Added condition to workflow status
- Set the condition in the controller. 

Would we like to set more conditions? Would we also like to

```
Deprecate the phase etc. fields in status which are superceded by Conditions
``` 
as suggested in the issue? 

Happy for feedback! 
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlike to be merged.
* [ ] Optional. I've added My organization is added to the USERS.md.
* [x] I've signed the CLA and required builds are green.

### CODE REVIEW

**Comment by NikeNano** [comment]
Will fix the tests!

**Comment by NikeNano** [comment]
I will look into tests as well.

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 1]
```diff
@@ -873,6 +873,15 @@ type WorkflowStatus struct {
 
 	// Outputs captures output values and artifact locations produced by the workflow via global outputs
 	Outputs *Outputs `json:"outputs,omitempty" protobuf:"bytes,8,opt,name=outputs"`
+
+	// Condition for k8s conditions https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties
+	Condition []ConditionStruct `json:"condition,omitempty" protobuf:"bytes,11,opt,name=condition"`
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/workflow_types.go]: Should this be `Conditions`?

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 2]
```diff
@@ -873,6 +873,15 @@ type WorkflowStatus struct {
...
+}
+
+type ConditionStruct struct {
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/workflow_types.go]: just `Condition` - I think struct in implied

File: pkg/apis/workflow/v1alpha1/workflow_types.go (hunk 1 above)
Comment (NikeNano) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: Is the naming important for 
```
kubectl
```
to be able to pick it up? If not let's go for ```Conditions```

Comment (alexec) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: I think it needs to be like this:

```
k get pod -o yaml
...
  status:
    conditions:
    - lastProbeTime: null
      lastTransitionTime: "2020-03-16T17:35:26Z"
      status: "True"
      type: Initialized
```

Comment (alexec) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: looking at the source code for corev1.PodStatus:
```
Conditions []PodCondition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,2,rep,name=conditions"`
```
```
// PodCondition contains details for the current condition of this pod.
type PodCondition struct {
	// Type is the type of the condition.
	// More info: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle#pod-conditions
	Type PodConditionType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=PodConditionType"`
	// Status is the status of the condition.
	// Can be True, False, Unknown.
	// More info: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle#pod-conditions
	Status ConditionStatus `json:"status" protobuf:"bytes,2,opt,name=status,casttype=ConditionStatus"`
	// Last time we probed the condition.
	// +optional
	LastProbeTime metav1.Time `json:"lastProbeTime,omitempty" protobuf:"bytes,3,opt,name=lastProbeTime"`
	// Last time the condition transitioned from one status to another.
	// +optional
	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty" protobuf:"bytes,4,opt,name=lastTransitionTime"`
	// Unique, one-word, CamelCase reason for the condition's last transition.
	// +optional
	Reason string `json:"reason,omitempty" protobuf:"bytes,5,opt,name=reason"`
	// Human-readable message indicating details about last transition.
	// +optional
	Message string `json:"message,omitempty" protobuf:"bytes,6,opt,name=message"`
}
```

Comment (alexec) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]:
```
type ConditionStatus string

// These are valid condition statuses. "ConditionTrue" means a resource is in the condition.
// "ConditionFalse" means a resource is not in the condition. "ConditionUnknown" means kubernetes
// can't decide if a resource is in the condition or not. In the future, we could add other
// intermediate conditions, e.g. ConditionDegraded.
const (
	ConditionTrue    ConditionStatus = "True"
	ConditionFalse   ConditionStatus = "False"
	ConditionUnknown ConditionStatus = "Unknown"
)
```

Comment (alexec) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: I do not think you want them all

Comment (NikeNano) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: Yeah according to the [doc](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties) ```but must contain at least type and status fields:``` I think type and status is enough for this? What are your thoughs? Update to ```Conditions```

**Comment by NikeNano** [comment]
Update and all should be good now @alexec :)

**Comment by alexec** [comment]
Delightfully simple solution. Thank you!

**Comment by jessesuen** [comment]
This feature is incomplete. When retrying a workflow, we do not reset the condition and we will have inconsistent state. The condition will remain true, whereas after retry the label is false (or unset, i forget which)

**Comment by NikeNano** [comment]
> plete. When retrying a workflow, we do not reset the condition and we will have inconsistent state. The condition will remain true, whereas after retry the label is false (or unset, i forget which)

Ohhh, I will look in to this to try to fix it!

**Comment by NikeNano** [comment]
@jessesuen would it be sufficient to update the workflow condition if we do a retry? If so I suggest we reset the condition [here](https://github.com/NikeNano/argo/blob/cd104a2b370a5813da69848a1ce7898d3876f316/workflow/controller/operator.go#L1338)

before retry node is called:

```
		processedRetryParentNode, continueExecution, err := woc.processNodeRetries(retryParentNode, *processedTmpl.RetryStrategy)
```

If it sounds good I will make a PR to fix it :)

**Comment by simster7** [comment]
@NikeNano I'm adding some major functionality to conditions as well as refactoring some of the code, so I'll cover this case for you :). No need to make a PR for this.

**Comment by NikeNano** [comment]
Thanks @simster7!

**Pull request merged**